src/object/cli_ec.c

```diff
@@ -2640,7 +2637,84 @@ obj_recx_ec2_daos(struct daos_oclass_attr *oca, int shard,
 		}
 	}
 
-	*output_nr = total;
-	*output_recxs = tgt_recxs;
+	D_FREE(*recxs_p);
+	*recxs_p = tgt_recxs;
+	*nr = total;
+	return 0;
+}
+
+/* Convert DAOS offset to specific data target daos offset */
+int
+obj_recx_ec_daos2shard(struct daos_oclass_attr *oca, int shard,
+		       daos_recx_t **recxs_p, unsigned int *iod_nr)
+{
+	daos_recx_t	*recx = *recxs_p;
+	int		nr = *iod_nr;
+	int		cell_nr = obj_ec_cell_rec_nr(oca);
+	int		stripe_nr = obj_ec_stripe_rec_nr(oca);
+	daos_recx_t	*tgt_recxs;
+	int		shard_idx = shard % obj_ec_tgt_nr(oca);
+	int		total;
+	int		idx;
+	int		i;
+
+	D_ASSERT(shard_idx < obj_ec_data_tgt_nr(oca));
+	for (i = 0, total = 0; i < nr; i++) {
+		uint64_t offset = recx[i].rx_idx;
+		uint64_t end = offset + recx[i].rx_nr;
+
+		while (offset < end) {
+			daos_off_t shard_start = rounddown(offset, stripe_nr) +
+						 shard * cell_nr;
+			daos_off_t shard_end = shard_start + cell_nr;
+
+			/* Intersect with the shard cell */
+			if (max(shard_start, offset) < min(shard_end, end))
+				total++;
+
+			offset = roundup(offset + 1, stripe_nr);
+		}
+	}
+
+	if (total == 0)
+		return 0;
+
+	D_ALLOC_ARRAY(tgt_recxs, total);
+	if (tgt_recxs == NULL)
+		return -DER_NOMEM;
+
+	for (i = 0, idx = 0; i < nr; i++) {
+		uint64_t offset = recx[i].rx_idx;
+		uint64_t end = offset + recx[i].rx_nr;
+
+		while (offset < end) {
+			daos_off_t shard_start = rounddown(offset, stripe_nr) +
+						 shard * cell_nr;
+			daos_off_t shard_end = shard_start + cell_nr;
+			uint64_t rx_idx;
+			uint64_t rx_nr;
+
+			if (max(shard_start, offset) >= min(shard_end, end)) {
+				offset = roundup(offset + 1, stripe_nr);
+				continue;
+			}
+
+			/* Intersect with the shard cell */
+			rx_idx = max(rounddown(offset, stripe_nr) +
+				     shard_idx * cell_nr, offset);
+			rx_nr = min(roundup(rx_idx + 1, cell_nr) - rx_idx,
+				    end - offset);
```

Comment:
seems possible incorrect
better to be "rx_nr = min(shard_end, end) - rx_idx"?

Reply:
yes, seems better.